src/meta/processors/partsMan/ListPartsProcessor.cpp

```diff
@@ -91,34 +96,28 @@ ListPartsProcessor::getAllParts() {
 
 
 void ListPartsProcessor::getLeaderDist(std::vector<cpp2::PartItem>& partItems) {
-    if (adminClient_ == nullptr) {
-        return;
-    }
-
-    HostLeaderMap hostLeaderMap;
-    auto ret = adminClient_->getLeaderDist(&hostLeaderMap).get();
-    if (!ret.ok() && !hostLeaderMap.empty()) {
-        LOG(ERROR) << "Get leader distribution failed";
+    const auto& hostPrefix = MetaServiceUtils::leaderPrefix();
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto kvRet = kvstore_->prefix(kDefaultSpaceId, kDefaultPartId, hostPrefix, &iter);
+    if (kvRet != kvstore::ResultCode::SUCCEEDED) {
         return;
     }
 
-    for (auto& hostEntry : hostLeaderMap) {
-        auto leader = hostEntry.first;
-        auto spaceIter = hostEntry.second.find(spaceId_);
-        if (spaceIter != hostEntry.second.end()) {
-            for (auto& partId : spaceIter->second) {
-                auto it = std::find_if(partItems.begin(), partItems.end(),
-                    [&] (const auto& partItem) {
-                        return partItem.part_id == partId;
-                    });
-                if (it != partItems.end()) {
-                    it->set_leader(this->toThriftHost(leader));
-                } else if (showAllParts_ || std::find(partIds_.begin(), partIds_.end(), partId) ==
-                                            partIds_.end()) {
-                    LOG(ERROR) << "Maybe not get the leader of partition " << partId;
+    auto activeHosts = ActiveHostsMan::getActiveHosts(kvstore_);
+    while (iter->valid()) {
+        auto host = MetaServiceUtils::parseLeaderKey(iter->key());
+        if (std::find(activeHosts.begin(), activeHosts.end(),
+                      HostAddr(host.ip, host.port)) != activeHosts.end()) {
+            LeaderParts leaderParts = MetaServiceUtils::parseLeaderVal(iter->val());
```

Comment:
each loop will create a template map, and will destruct after the loop, maybe here can define a function named parseLeaderValBySpaceID(iter->val(), spaceId_) to just return the leaders of current space.

Reply:
Good suggestion